### PR TITLE
Add Offline Analysis Tool for Mobile Web MCP

### PR DESCRIPTION
**Pull request by haifeng-li-at-salesforce**
# 🚀 Add Offline Analysis Tool for Mobile Web MCP

## 📋 Summary

This PR introduces a comprehensive offline analysis tool for Salesforce Lightning Web Components (LWC) that provides mobile-specific code analysis and recommendations. The tool integrates with the Model Context Protocol (MCP) to offer expert guidance for mobile-ready LWC development.

### 🔧 Core Offline Analysis Tool
- **Mobile Web Offline Analysis Tool** (`OfflineAnalysisTool`)
  - Analyzes LWC components for mobile-specific issues
  - Provides detailed recommendations for mobile readiness
  - Integrates with ESLint and LWC Graph Analyzer plugin
  - Supports comprehensive error handling and graceful degradation

### 📊 Analysis Capabilities
- **Rule-based Analysis**: Implements configurable rules for mobile-specific issues
- **Wire Configuration Analysis**: Detects private wire configuration properties
- **Reactive Value Analysis**: Identifies non-local property references in wire configs
- **Code Quality Assessment**: Provides actionable recommendations for improvements

### 🧪 Comprehensive Test Coverage
- **Unit Tests**: Complete test suite for all tool functionality
- **Error Handling Tests**: Robust error scenario coverage
- **Integration Tests**: End-to-end tool registration and analysis testing
- **Test Helpers**: Reusable test utilities for consistent testing patterns

### New Tool Structure
```
packages/mobile-web/src/tools/mobile-offline/offline-analysis/
├── tool.ts              # Main OfflineAnalysisTool implementation
└── ruleConfig.ts        # Configurable analysis rules
```

### Enhanced Base Tool
- Updated `BaseTool` signature for better tool registration
- Improved error handling and type safety
- Enhanced service grounding text generation

### Schema Enhancements
- **Analysis Schema**: New schemas for code analysis results
- **LWC Schema**: Enhanced LWC code type definitions
- **Tool Interface**: Improved tool registration interface

## Sample output
image

### Test Coverage
- ✅ **12/12 tests passing** for the offline analysis tool
- ✅ **Comprehensive error handling** test scenarios
- ✅ **Tool registration** and integration tests
- ✅ **Code analysis** functionality tests
- ✅ **Rule configuration** validation tests

### Test Categories
1. **Tool Properties**: Validates tool metadata and configuration
2. **Tool Registration**: Tests MCP server integration
3. **Code Analysis**: End-to-end analysis functionality
4. **Error Handling**: Robust error scenario coverage
5. **Private Methods**: Internal functionality validation
6. **Tool Integration**: Compatibility with existing patterns

## ✅ Checklist

- [x] Core offline analysis tool implementation
- [x] Comprehensive test coverage (12/12 tests passing)
- [x] Error handling and edge case coverage
- [x] MCP integration and tool registration
- [x] Rule configuration system
- [x] Type definitions and schemas
- [x] Documentation and usage examples
- [x] Performance optimization
- [x] Code quality and linting compliance

## 🎯 Impact

This PR significantly enhances the mobile web MCP tools by providing:
- **Expert Analysis**: Professional-grade LWC analysis
- **Mobile Optimization**: Specialized mobile development guidance
- **Developer Experience**: Improved development workflow
- **Quality Assurance**: Automated code quality checks
- **Scalability**: Extensible rule-based analysis system

---

**Ready for Review** ✅  
**Tests Passing** ✅  
**Documentation Complete** ✅  
**Performance Optimized** ✅

### CODE REVIEW

File: packages/mobile-web/src/schemas/analysisSchema.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import { z } from 'zod';
```
Comment (ben-zhang-at-salesforce) [inline packages/mobile-web/src/schemas/analysisSchema.ts:1]: missing copyrights chunk

File: packages/mobile-web/src/schemas/analysisSchema.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    .describe(
+      `How a developer should address the issue? Be as detailed as possible without adding code snippets.`
+    ),
+});
+
+export const CodeAnalysisIssueSchema = CodeAnalysisBaseIssueSchema.extend({
+  code: z.string().optional().describe('What is the code snippet with the issue?'),
+  location: z
+    .object({
+      startLine: z.number(),
```
Comment (ben-zhang-at-salesforce) [inline packages/mobile-web/src/schemas/analysisSchema.ts:24]:
```suggestion
      startLine: z.number().gte(0),
``` ?

File: packages/mobile-web/src/schemas/lwcSchema.ts
Comment (ben-zhang-at-salesforce) [inline packages/mobile-web/src/schemas/lwcSchema.ts:1]: same missing copyright chunk

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+  private getOrchestrationInstructions(): string {
+    return "If you have not already executed the 'sfmobile-web-offline-guidance' tool in this conversation thread: (1) Immediately call the 'sfmobile-web-offline-guidance' tool, (2) Execute all review instructions from the guidance tool's output against the user's LWC component code in the specified format, (3) Combine the review results from step 2 with this tool's analysis results, then refactor the component based on all combined review recommendations.";
+  }
+
+  private createAnalysisMarkdown(
+    componentName: string,
+    analysisResults: ExpertsCodeAnalysisIssuesType
+  ): string {
+    let markdown = `Analysis results for the LWC component ${componentName} are presented below. Please review them and provide a detailed refactoring plan for the component. \n\n`;
```
Comment (ben-zhang-at-salesforce) [inline packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:144]: nitpick:   string concatenation is expensive.  use array to hold the individual, by the end do a array.join() to get the final result?

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+    for (const analysis of analysisResults.analysisResults) {
+      markdown += this.formatAnalysisSection(analysis);
```
Comment (ben-zhang-at-salesforce) [inline packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:147]: if not issue found calling formatAnalysisSection(),  return undefined and skip appending it the markdown?

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+    for (const issue of analysis.issues) {
+      section += this.formatIssue(issue);
+    }
+
+    return section;
+  }
+
+  private formatIssue(issue: CodeAnalysisIssueType): string {
+    let issueText = `## ${issue.type}\n\n`;
```
Comment (ben-zhang-at-salesforce) [inline packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:173]: avoid string concatenation?

**Review by jhorst11 — APPROVED** [review]
can we fix the prettier formatting - otherwise besides ben's feedback LGTM - I'll need this in to get going on the offline-guidance too to share some of these new types and schemas

File: packages/mobile-web/resources/appReview/appReviewService.d.ts [hunk 6]
```diff
@@ -4,7 +4,7 @@
  * For full license text, see the LICENSE.txt file
  */
 
-import { BaseCapability } from "../BaseCapability.js";
+import { BaseCapability } from '../BaseCapability.js';
```
[the remark this answers is not in the document]
Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/resources/appReview/appReviewService.d.ts:7]: Good point! The prettierignore file at the root includes the resources folder. After re-fetching the resource TypeScript files and re-running Prettier, everything looks good.

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+          return {
+            content: [
+              {
+                type: 'text',
+                text: this.createAnalysisMarkdown(code.name, analysisResults),
+              },
+            ],
+          };
+        } catch (error) {
+          console.error('Error during code analysis:', error);
```
[the remark this answers is not in the document]
Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:103]: Good point! Updated.

**Review by ben-zhang-at-salesforce — APPROVED** [review]
looks good!

File: packages/mobile-web/src/index.ts [hunk 8]
```diff
@@ -37,30 +38,22 @@ const annotations: ToolAnnotations = {
...
+  new CalendarTool(),
+  new ContactsTool(),
+  new DocumentScannerTool(),
+  new GeofencingTool(),
+  new LocationTool(),
+  new NfcTool(),
+  new PaymentsTool(),
+  new OfflineAnalysisTool(),
+];
+// Register all tools
```
Comment (khawkins) [inline packages/mobile-web/src/index.ts:55]:
```suggestion

// Register all tools
```

File: packages/mobile-web/src/schemas/lwcSchema.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import z from 'zod';
+
+import { zodToJsonSchema } from 'zod-to-json-schema';
+
+export const EmptySchema = z.object({});
+
+export const EmptyJsonSchema = zodToJsonSchema(EmptySchema);
```
Comment (khawkins) [inline packages/mobile-web/src/schemas/lwcSchema.ts:13]: Are we anticipating the use of this? It's not currently used.

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/ruleConfig.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2025, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { CodeAnalysisBaseIssueType } from '../../../schemas/analysisSchema';
```
Comment (khawkins) [inline packages/mobile-web/src/tools/mobile-offline/offline-analysis/ruleConfig.ts:7]: I would like us to standardize on explicitly adding the `.js` extension to our import statements, for our local resources. I would actually like to ensure that we are a `type: "module"` (ESM) package, because there's really no reason to carry around the legacy of CommonJS in this day and age, and lots of reasons to abandon it. But I'm not sure we need to make that task a part of this PR.

At a minimum though, I want to be prepared with the proper ESM import statements with `.js` extensions, for our code imports.

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+  CodeAnalysisIssueType,
+  ExpertsCodeAnalysisIssuesSchema,
+  ExpertsCodeAnalysisIssuesType,
+  ExpertCodeAnalysisIssuesType,
+} from '../../../schemas/analysisSchema';
+
+const BUNDLE_ANALYZER = lwcGraphAnalyzerPlugin.processors.bundleAnalyzer;
+const ANALYSIS_EXPERT_NAME = 'Mobile Web Offline Analysis';
+const PLUGIN_NAME = '@salesforce/lwc-graph-analyzer';
+const RECOMMENDED_CONFIG = lwcGraphAnalyzerPlugin.configs.recommended;
```
[the remark this answers is not in the document]
Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:27]: Very good point!

File: packages/mobile-web/package.json [hunk 12]
```diff
@@ -20,7 +20,10 @@
     "update-resources": "tsx src/scripts/update-type-declarations.ts"
   },
   "dependencies": {
-    "@modelcontextprotocol/sdk": "^1.13.2"
+    "@modelcontextprotocol/sdk": "^1.13.2",
```
[the remark this answers is not in the document]
Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/package.json:23]: Done

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+    this.linter = new Linter({ configType: 'flat' });
+    this.ruleReviewers = this.initializeRuleReviewers();
+  }
+
+  public register(server: McpServer, annotations: ToolAnnotations): void {
+    server.registerTool(
+      this.toolId,
+      {
+        description: this.description,
+        inputSchema: this.inputSchema.shape,
```
Comment (khawkins) [inline packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:94]: We should be setting `outputSchema` in here as well.

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+              {
+                type: 'text',
+                text: this.createAnalysisMarkdown(code.name, analysisResults),
+              },
+            ],
+          };
+        } catch (error) {
+          return {
+            isError: true,
+            content: [
```
Comment (khawkins) [inline packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:112]: And we should be additionally sending `structuredContent` with the filled-in content for the `ExpertsCodeAnalysisIssuesSchema` schema. See https://modelcontextprotocol.io/specification/draft/server/tools#output-schema.

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts (hunk 13 above)
Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:94]: That was the approach I tried earlier, but it didn’t work. When outputSchema is set, the MCP host cannot interpret the output JSON and fails to produce the expected results. It’s said that the MCP SDK needs to parse the output JSON. To help the AI agent understand the MCP response, the JSON object is converted to a markdown file and returned as plain text.

The below is the code that failed earlier:
image

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts (hunk 14 above)
Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:112]: The `structuredContent` is not comprehended by AI agent. I tried the approach to ask the AI agent to reference the returned JSON object in the `content` section, it didn't work.

File: packages/mobile-web/src/schemas/lwcSchema.ts (hunk 9 above)
Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/src/schemas/lwcSchema.ts]: removed

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts (hunk 13 above)
Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:56]: Yes, the inspector correctly receives the structuredContent as expected, which can be verified in the UI. The issue is that the AI agent (Cursor, in this case) does not process structured data and therefore does not use it during code review. The structuredContent in the MCP server tool response is meant for the MCP SDK, but apparently not for agents like Cursor.

Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:56]: The screenshot below shows the Inspector behavior, where the structured content is correctly recognized and displayed:
image

In contrast, the screenshot below shows the Cursor behavior, which only uses the unstructured data, ignores the structured content, and returns SLDS-related noise:
image

Comment (khawkins) [reply packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:56]: > The structuredContent in the MCP server tool response is meant for the MCP SDK, but apparently not for agents like Cursor.

- `structuredContent` is the direction that the MCP specification is headed. It provides a more deterministic and, well, structured dataset to the MCP client, which will in turn be receiving better context to solicit its response.
- Cursor will make use of `structuredContent` when properly formed.
- When using `structuredContent`, per the [spec](https://modelcontextprotocol.io/specification/draft/server/tools#structured-content) it's *still required* that you continue to *also* provide `content`, for backward compatibility with *older* MCP clients. *But*: with notable direction:

    > For backwards compatibility, a tool that returns structured content SHOULD also return **functionally equivalent** unstructured content. (For example, **serialized JSON** can be returned in a TextContent block.) [emphasis mine]

    The issue is called out in MCP Inspector, in the warning you receive about your (unstructured) `content`:

    > Unstructured content is not valid JSON

- Take a look at my code example [above](https://github.com/forcedotcom/mobile-mcp-tools/pull/21#discussion_r2191221917):
    - I'm simply returning `analysisResults` as the `structuredContent` value, because it conforms to the output schema.
    - I'm returning `JSON.stringify(analysisResults)` as the `content` value, i.e. *functionally equivalent* to the `structuredContent` value, provided for backwards compatibility.

Basically, at some point `structuredContent` will be the *expected* format for return values of MCP server tools, except for edge cases. `content` is meant to serve as a point of backwards compatibility for older MCP clients.

I packaged up my code changes and ran them in a consuming project, and this is what Cursor gives me:

image

While it could probably use some more directed rationale tweaking for its output, it does the right thing and makes the changes to the component.

Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:56]: Nice finding! The change is made.

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+      (acc, ruleConfig) => {
+        acc[ruleConfig.id] = ruleConfig.config;
+        return acc;
+      },
+      {} as Record<string, CodeAnalysisBaseIssueType>
+    );
+  }
+
+  private async analyzeCode(code: LwcCodeType): Promise<ExpertsCodeAnalysisIssuesType> {
+    const jsCode = code.js.map(js => js.content).join('\n');
```
Comment (khawkins) [inline packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:99]: This has got me realizing that I don't think our schema / output is complete. My first thought here was: "We can't concatenate code together because there can be issues even with the fact that there are multiple files at all. For example, Komaci does not allow for multiple template files ([`no-multiple-template-files`](https://github.com/salesforce/eslint-plugin-lwc-graph-analyzer/blob/37ec92c25abb0be9883f5de66f309e46a0e054d3/lib/docs/no-multiple-template-files.md))."

But then I realized that, in our schemas, we don't actually provide a tie-back from the input code to the output issues. This should probably be in the `ExpertCodeAnalysisIssuesSchema` type, and then we'd run analysis on each file within the bundle, tying it back to the code's path.

Anyway, let's keep it as is for this PR, and I'll think about what the revision should look like.

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts (hunk 14 above)
Comment (khawkins) [reply packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:76]: Ah yes, I hadn't even realized that I added this comment on the error case. Now our discussion makes more sense! 😆 I had thought I commented on the non-error case above.

**Review by khawkins — APPROVED** [review]
Looks great! Thanks for your efforts. I'll follow up on the schema updates and some odds and ends.

File: packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts (hunk 15 above)
Comment (haifeng-li-at-salesforce) [reply packages/mobile-web/src/tools/mobile-offline/offline-analysis/tool.ts:99]: All right, we will address it in the next PR.
